[ParseableInterfaces] Refine cache keys and dependency checks

harlanhaskins opened the pull request:
Hashing the contents of the interface files for dependency checking is overkill. In practice, size and last modification time are enough to determine if a file has changed on disk, and therefore should be rebuilt.

As for cache keys, previously, we included the PCH hash components in the cache key. While they didn’t do any harm, they didn’t contribute any unique information about the module in question.

Additionally, passing the effective language version in means that each dependency that uses a different -swift-version would re-compile all of its dependencies. This is unfortunate, as that means the standard library is recompiled potentially several times.

rdar://46503065

harlanhaskins committed:
[Serialization] Use mtime instead of hash for file dependencies

Hashing the contents of the interface files is overkill. In practice, size and last modification time are enough to determine if a file has changed on disk, and therefore should be rebuilt.

harlanhaskins committed:
[ParseableInterface] Refine cache key for cached swiftmodules

Previously, we included the PCH hash components in the cache key. While they didn’t do any harm, they didn’t contribute any unique information about the module in question.

Additionally, passing the effective language version in means that each dependency that uses a different -swift-version would re-compile all of its dependencies. This is unfortunate, as that means the standard library is recompiled potentially several times.

harlanhaskins committed:
[test] Change hash-based module cache tests to mtime-based

These tests made sure that changing the hash, while keeping mtime constant, causes dependent modules to be rebuilt. Flip this around so changing the mtime, while keeping the hash constant, causes the same rebuilding behavior.

jrose-apple commented on test/ParseableInterface/ModuleCache/module-cache-intermediate-mtime-change-rebuilds-only-intermediate.swift:
```diff
@@ -27,12 +27,11 @@
 //
 //
-// Actual test: Change a byte in OtherModule.swiftinterface, check we only rebuild its cached module.
+// Actual test: Change the mtime of OtherModule.swiftinterface, check we only rebuild its cached module.
```
Are there already tests for changing the size but keeping the mtime constant?

harlanhaskins replied:
Yep, for both the leaf and intermediate modules.

jrose-apple commented on lib/Frontend/ParseableInterfaceSupport.cpp:
```diff
@@ -124,20 +125,23 @@ static std::string getCacheHash(ASTContext &Ctx,
+  // Explicitly don't pass in the "effective" language version -- this would
+  // mean modules built in different -swift-version modes would rebuild their
+  // dependencies.
+  llvm::hash_code H = hash_value(swift::version::getSwiftFullVersion());
```
Can you add a test that this doesn't happen anymore?

harlanhaskins replied:
Ah! Yes, thanks for the reminder.

harlanhaskins replied:
Added the test. Fails with an older build, succeeds with this one.

harlanhaskins committed:
[test] Make sure we don't double-compile parseable dependencies

Adds a test to make sure that there's only one version of a dependent
module, regardless how many -swift-versions are present in the
dependency hierarchy.
